### PR TITLE
Support running tests in full VM's

### PR DESCRIPTION
**Pull request by adam-stokes**
- Orchestrates deploying pure vm's for both amd64/arm64, utilizing ansible. Windows support coming next in followup
- Provides uniqueness with policies and integrations so that all test suites can run against a single deployed stack
- Update the step definitions to be a little more platform agnostic, still more work to do there in followup pr's

Signed-off-by: Adam Stokes <51892+adam-stokes@users.noreply.github.com>

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have run the Unit tests (`make unit-test`), and they are passing locally
- [x] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [x] I have noticed new Go dependencies (run `make notice` in the proper directory)

## How to test this PR locally

Please see `.ci/README.md`

## Related issues

- Closes https://github.com/elastic/observability-robots/issues/720

### CODE REVIEW

File: .ci/.e2e-tests.yaml
Comment (mdelapenya) [inline .ci/.e2e-tests.yaml:23]: Should we use an array struct here?

```yaml
- labels:
   - "arm64"
   - "amd64"
```

Comment (mdelapenya) [reply .ci/.e2e-tests.yaml:23]: I think we should replicate changes in this file in the `.ci/.e2e-tests-for-elastic-agent.yaml` file

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:325]: I think this is more correct 🤔 
```suggestion
        if(suite == "fleet" && (platformLabels == "arm64" || platformLabels == "amd64")){
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:267]: Another candidate for the library?

Comment (mdelapenya) [inline .ci/Jenkinsfile:267]: Wdyt? Because we are passing command AND args 🤔 
```suggestion
def pyrun(command){
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:287]: Because of https://github.com/elastic/e2e-testing/pull/1740/files#r741322752
```suggestion
            ${env.WORKSPACE}/.venv/bin/${command}
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:308]: Wdyt using a more descriptive tag? IIUC, this is for provisioning the stack before the tests run, right? 
```suggestion
    ansible("-t provision-stack")
```
We should update ansible tags in consequence

File: .ci/ansible/roles/common/tasks/setup_test_script.yml
Comment (mdelapenya) [inline .ci/ansible/roles/common/tasks/setup_test_script.yml:21]: We already have a script to run the tests in the same path. Do we want to use the old one or we don't care about overriding it?

Comment (adam-stokes) [reply .ci/ansible/roles/common/tasks/setup_test_script.yml:21]: I trimmed this one down for VM runs only, at least until I get green across the board. It's mainly to help with debugging atm

File: .ci/Jenkinsfile
Comment (adam-stokes) [reply .ci/Jenkinsfile:287]: The latest code looks to be using `args` here

File: internal/deploy/remote.go
Comment (mdelapenya) [inline internal/deploy/remote.go:67]: For the future, we may need to consume a library like GoSysinfo (https://github.com/elastic/go-sysinfo/blob/2b7239d64d62e2e26f626a71b4730dd78086d3f8/types/host.go#L67)

File: internal/elasticsearch/client.go
Comment (mdelapenya) [inline internal/elasticsearch/client.go:288]: Should we update the reference to the elasticsearch URL here too?

File: internal/kibana/fleet.go
Comment (mdelapenya) [inline internal/kibana/fleet.go:55]: Ditto: is the URL retrieved from the environment?

File: internal/utils/utils.go
Comment (mdelapenya) [inline internal/utils/utils.go:31]: Should we read from the env first or directly get the value from runtime?

File: internal/elasticsearch/client.go
Comment (adam-stokes) [reply internal/elasticsearch/client.go:288]: Yes! Good catch

File: internal/kibana/fleet.go
Comment (adam-stokes) [reply internal/kibana/fleet.go:55]: Correct again! I'll update this

File: internal/utils/utils.go
Comment (adam-stokes) [reply internal/utils/utils.go:31]: So for pure vm's we can pull from the runtime, for things like autodiscover which uses docker we may need to still pull that env var?

Comment (mdelapenya) [reply internal/utils/utils.go:31]: I'm not sure if we even use this method elsewhere than fleet 🤔

File: internal/deploy/remote.go
Comment (adam-stokes) [reply internal/deploy/remote.go:67]: Right now working with https://github.com/shirou/gopsutil

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:10]:
```suggestion
export RUN_ID=$(uuidgen|cut -d'-' -f1)
```

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:1]: I'm OK having this doc here in the `.ci` dir, although I'd prefer having it in a place developers are able to find it in an easy manner, maybe creating a dedicated section `How to run the tests` in the main Readme. Nevertheless, this can be done in a follow up PR, given this one is intended to run on CI only

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:329]: I guess this is going to spread the test workloads into one host, defined by `scenarioIdx`:

First scenario to the first host, and so on, right? If that's the case, we will be creating as many hosts as items in the YAML descriptor

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:466]: Maybe we can extract `${testRunner.split('@')[0]}` and `${testRunner.split('@')[1]}` to variables to simplify the code and increase readability

```groovy
def sshUser = ${testRunner.split('@')[0]}
def sshHostIP = ${testRunner.split('@')[1]}
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:455]: We must make sure the OTEL variables are sent to the testRunner machine, as I guess this wrapper won't work for the remote SSH connection. Let's think about the security implications of sending those credentials to the machines 🤔 

With that in mind, if we are not sending APM traces it could be OK until we resolve this concern

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:470]: Could we extract this block into a function? Something like `localRunE2E` or any other name you see convenient. It would help reading this Jenkinsfile

**Review by mdelapenya — APPROVED** [review]
I think this PR is great Adam, thanks for this effort! I added some minor comments, many of them related to follow-up work that I did not want to forget for future reference. Awesome!!!

image

File: .ci/Jenkinsfile
Comment (adam-stokes) [reply .ci/Jenkinsfile]: Just to elaborate, one of the ideas is to make vault accessible from within those test runners to get the necessary info to send to APM

**Comment by cachedout** [comment]
@adam-stokes Where are we with this? It look like it needs a rebase and there are still some failing tests.

We have the following issues which are blocked on this:

https://github.com/elastic/observability-robots/issues/720
https://github.com/elastic/e2e-testing/issues/1658
https://github.com/elastic/e2e-testing/issues/1340
https://github.com/elastic/e2e-testing/issues/1511

**Comment by adam-stokes** [comment]
@cachedout I have everything running and passing locally, just tweaking jenkinsfile to run it all now

**Comment by adam-stokes** [comment]
Just an update, we are down to fixing final flakiness in the background_processes test. I've run these tests locally multiple times which mimics exactly what CI does and the tests are passing, however, as soon as CI attempts it the background_processes tests fail to bring up the second remaining filebeat service on agent restart

**Comment by cachedout** [comment]
>  I've run these tests locally multiple times which mimics exactly what CI does and the tests are passing

Perhaps you could [launch a worker instance](https://github.com/elastic/infra/tree/master/docs/ci#how-can-i-launch-an-instance) and see if you can replicate the issue there?

**Comment by adam-stokes** [comment]
Everything is now running inside their own nodes which then provisions the test nodes for our test suites.  From this point everything should be as isolated as you can get and should hopefully provide additional stability in running the tests.

**Comment by cachedout** [comment]
@adam-stokes Could we please have another status update for this PR? Thanks!

**Comment by mdelapenya** [comment]
> @adam-stokes Could we please have another status update for this PR? Thanks!

Wdyt about merging this PR without the run-on-tops-of-beats feature file (or those that are consistently failing)? The we will be adding value with the VM support meanwhile we work on the permanent fix or root cause analysis of the issue with backend processes.

**Comment by adam-stokes** [comment]
@mdelapenya that works for me, the actual orchestrating of everything is complete and ready to go, now im just focusing on the tests. I'll comment out the testing that are failing then propose this for merging

File: .ci/.e2e-tests-for-elastic-agent.yaml
Comment (mdelapenya) [inline .ci/.e2e-tests-for-elastic-agent.yaml:27]: It's weird that we reference a MACHINES[key] with `platforms`. Maybe `machines`? or rename the top object to `PLATFORMS`?

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:221]: If we are in this else branch, will the next stage (End-To-End Tests) be run? If the else branch is evaluated (both FORCE_SKIP_GIT_CHECKS and isGitRegionMatch are false), then no stack will be provisioned

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:451]: I'd say so, as we have removed the `pullRequestFilter` attribute from the YAML file. What I'm not sure if we want to still keep that ability. In the past, we used it to separate Centos executions from Debian ones. Now that we run on VMs, maybe it's time to retire that ability

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:495]: More readable :)
```suggestion
            def testRunner = getNodeIp("${env.WORKSPACE}", platform)
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:509]: More readable :)
```suggestion
        def testRunner = getNodeIp("${env.WORKSPACE}", platform)
```

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:526]: For clarity, let's use parenthesis for multiline steps
```suggestion
        junit(allowEmptyResults: true,
          keepLongStdio: true,
          testResults: "outputs/${testRunner.ip}/TEST-*${runId}*.xml")
        archiveArtifacts(allowEmptyArchive: true,
          artifacts: "outputs/${testRunner.ip}/TEST-*${runId}*.xml")
```

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:29]: Let's explain where the `nodeLabel`, `nodeImage` and `nodeInstanceType` comes from, or at least show a list of possible values

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:42]: Let's explain where the `nodeLabel`, `nodeImage` and `nodeInstanceType` comes from, or at least show a list of possible values

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:56]: Let's explain where the `nodeLabel`, `nodeImage` and `nodeInstanceType` comes from, or at least show a list of possible values

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:69]: Let's explain where the `nodeLabel`, `nodeImage` and `nodeInstanceType` comes from, or at least show a list of possible values

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:82]: Let's explain that different test suites can be used passing Gherkin tags. How would we select a different test suite, k8s-autodiscover or helm?

File: .ci/README.md
Comment (mdelapenya) [inline .ci/README.md:20]: Should we add them to a requirements.txt file and install from there?

File: e2e/_suites/fleet/fleet.go
Comment (mdelapenya) [inline e2e/_suites/fleet/fleet.go:133]: Indeed. Let's make sure the method is not used anymore and deleted in consequence
```suggestion
```

File: e2e/_suites/fleet/fleet.go
Comment (mdelapenya) [inline e2e/_suites/fleet/fleet.go:207]: This is candidate to extract to a method, so that it's easier to be read

File: internal/action/process.go
Comment (mdelapenya) [inline internal/action/process.go:16]: Let's not forget to notice this library, and check license compatibility

File: internal/action/process.go
Comment (mdelapenya) [inline internal/action/process.go:6]: BTW, I loved the refactor using a library for process management 👏

File: internal/installer/elasticagent_tar.go
Comment (mdelapenya) [inline internal/installer/elasticagent_tar.go:52]: I do not know how many times we have done this change, and reverted it 😄 Hopefully the team will be consistent in the future

File: internal/kibana/fleet.go
Comment (mdelapenya) [inline internal/kibana/fleet.go:57]: Should we embed the credentials in the Elasticsearch Endpoint struct, to avoid reading the password multiple times?

**Review by mdelapenya — APPROVED** [review]
@adam-stokes great job in this PR! Fabulous how you dealt with infra provisioning, Jenkins orchestrating, Go programming, Gherkin refactors... Humongous thanks!

I added a few comments, minor, but please take a look at them before merging. They are basically related to readability

File: e2e/_suites/fleet/fleet.go
Comment (adam-stokes) [reply e2e/_suites/fleet/fleet.go:207]: Agreed, in a follow-up I wanted to support creating policy with optional integration added

Comment (adam-stokes) [reply e2e/_suites/fleet/fleet.go:207]: https://github.com/elastic/e2e-testing/issues/1853

**Pull request merged**
